File: jure-drf/notifications/models/notification.py
from django.conf import settings
from django.db import models


class Notification(models.Model):
    """
    Represents a single notification for a specific user.
    Delivered via in-app WebSocket and/or email depending
    on the notification type and user preferences.
    """

    recipient = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="notifications",
    )

    notification_type = models.CharField(max_length=60, db_index=True)
    title = models.CharField(max_length=200)
    message = models.TextField()

    related_case = models.ForeignKey(
        "cases.Case",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="notifications",
    )
    related_task = models.ForeignKey(
        "tasks.Task",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="notifications",
    )
    related_appointment = models.ForeignKey(
        "tasks.Appointment",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="notifications",
    )
    related_user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="triggered_notifications",
    )

    action_url = models.CharField(max_length=500, blank=True)

    is_read = models.BooleanField(default=False)
    read_at = models.DateTimeField(null=True, blank=True)

    priority = models.CharField(
        max_length=20,
        choices=[
            ("LOW", "LOW"),
            ("MEDIUM", "MEDIUM"),
            ("HIGH", "HIGH"),
            ("URGENT", "URGENT"),
        ],
        default="MEDIUM",
    )

    email_sent = models.BooleanField(default=False)
    email_sent_at = models.DateTimeField(null=True, blank=True)
    push_sent = models.BooleanField(default=False)

    created_at = models.DateTimeField(auto_now_add=True)
    expires_at = models.DateTimeField(null=True, blank=True)

    class Meta:
        ordering = ["-created_at"]
        indexes = [
            models.Index(fields=["recipient", "is_read"]),
            models.Index(fields=["recipient", "created_at"]),
            models.Index(fields=["notification_type", "created_at"]),
        ]

    def __str__(self):
        return f"{self.notification_type} → {self.recipient_id}"

File: jure-frontend/src/types/notification.ts
export type NotificationPriority = 'URGENT' | 'HIGH' | 'MEDIUM' | 'LOW';

export interface RelatedCase {
  id: number;
  reference?: string | null;
  title?: string | null;
  case_type?: string | null;
  caseType?: string | null;
}

export interface RelatedTask {
  id: number;
  title?: string | null;
}

export interface RelatedAppointment {
  id: number;
  title?: string | null;
}

export interface RelatedUser {
  id: number;
  firstName?: string | null;
  lastName?: string | null;
  /** Django / snake_case */
  first_name?: string | null;
  last_name?: string | null;
}

/** Normalized from GET /api/v1/notifications/ and WebSocket payloads. */
export interface AppNotification {
  id: number | string;
  type: string;
  title: string;
  message: string;
  priority?: NotificationPriority | string;
  is_read: boolean;
  read_at?: string | null;
  created_at: string;
  expires_at?: string | null;
  email_sent?: boolean;
  push_sent?: boolean;
  /** Relative path e.g. /dashboard/cases?case=ABC123 */
  action_url?: string | null;
  related_case?: RelatedCase | null;
  related_task?: RelatedTask | null;
  related_appointment?: RelatedAppointment | null;
  related_user?: RelatedUser | null;
  /** Derived label for UI chips */
  context_label?: string | null;
  /** Legacy flat fields (still supported if API omits nested) */
  case_reference?: string | null;
  case_title?: string | null;
  task_title?: string | null;
}

export type NotificationFilterId =
  | 'all'
  | 'unread'
  | 'urgent'
  | 'cases'
  | 'tasks'
  | 'appointments'
  | 'finance'
  | 'team'
  | 'messages';

/** localStorage preferences (V2: backend). */
export interface NotificationPrefs {
  tasks: boolean;
  cases: boolean;
  appointments: boolean;
  messages: boolean;
  finance: boolean;
  team: boolean;
  email: boolean;
}
